examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2926 @@
+#include <boost/range/iterator_range.hpp>
+
+// @sect3{Class template declarations}
+//
+// We begin our actual implementation by declaring all classes with its
+// data structures and methods upfront. In contrast to previous example
+// steps we use a more fine-grained encapsulation of concepts, data
+// structures, and parameters into individual classes. A single class thus
+// usually centers around either a  single data structure (such as the
+// Triangulation) in the <code>Discretization</code> class, or a single
+// method (such as the <code>step()</code> function of the
+// <code>TimeStep</code> class). We typically declare parameter variables
+// and scratch data object private and make methods and data structures
+// used by other classes public.
+//
+// @note A cleaner approach would be to guard access to all data
+// structures by <a
+// href="https://en.wikipedia.org/wiki/Mutator_method">getter/setter
+// functions</a>. For the sake of brevity, we refrain from that approach,
+// though.
+//
+// We also note that the vast majority of classes is derived from
+// ParameterAcceptor. This facilitates the population of all the global
+// parameters into a single (global) ParameterHandler. More explanations
+// about the use inheritance from ParameterAcceptor as a global subscription
+// mechanism can be found in Step-59.
+
+namespace Step69
+{
+  using namespace dealii;
+
+  // We start with an enum describing all possible boundary conditions
+  // encountered in this tutorial step. Such an enum allows us to refer to
+  // boundary types by a mnemonic (such as
+  // <code>Boundary::do_nothing</code>) rather than a numerical value.
+  enum Boundary : types::boundary_id
+  {
+    do_nothing = 0,
+    slip       = 1,
+    dirichlet  = 2,
+  };
+
+  // @sect4{The <code>Discretization</code> class}
+  //
+  // The class <code>Discretization</code> contains all data structures
+  // concerning the mesh (triangulation) and discretization (mapping,
+  // finite element, quadrature) of the problem. As mentioned, we use
+  // ParameterAcceptor class to automatically populate problem-specific
+  // parameters, such as the geometry information
+  // (<code>length</code>, etc.) or the refinement level
+  // (<code>refinement</code>) from a parameter file. This requires us to
+  // split the initialization of data structures into two functions: We
+  // initialize everything that does not depend on parameters in the
+  // constructor, and defer the creation of the mesh to the
+  // <code>setup()</code> method that can be called once all parameters are
+  // read-in via ParameterAcceptor::initialize().
+
+  template <int dim>
+  class Discretization : public ParameterAcceptor
+  {
+  public:
+    Discretization(const MPI_Comm &   mpi_communicator,
+                   TimerOutput &      computing_timer,
+                   const std::string &subsection = "Discretization");
+
+    void setup();
+
+    const MPI_Comm &mpi_communicator;
```

Comment:
We've generally just used copies of the `MPI_Comm`, rather than references to it. I would suggest doing the same here, and also in the constructor. There are likely more places elsewhere in the code where you have `MPI_Comm *\&`.

Reply:
done